pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,449 @@
+
+            # if the number of shots is 0, do nothing
+            if s == 0:
+                continue
+
+            # set the QNode device shots
+            h.device.shots = [(1, s)]
+
+            jacs = []
+            for i in argnums:
+                j = qml.jacobian(h, argnum=i)(*args, **kwargs)
+
+                if s == 1:
+                    j = np.expand_dims(j, 0)
+
+                # Divide each term by the probability per shot. This is
+                # because we are sampling one at a time.
+                jacs.append(c * j / p)
+
+            grads.append(jacs)
+
+        return [np.concatenate(i) for i in zip(*grads)]
+
+    @staticmethod
+    def check_device(dev):
+        r"""Verifies that the device used by the objective function is non-analytic.
+
+        Args:
+            dev (.Device): the device to verify
+
+        Raises:
+            ValueError: if the device is analytic
+        """
+        if dev.analytic:
+            raise ValueError("The Rosalin optimizer can only be used with non-analytic devices")
+
+    def check_learning_rate(self, coeffs):
+        r"""Verifies that the learning rate is less than 2 over the Lipschitz constant,
+        where the Lipschitz constant is given by :math:`\sum |c_i|` for Hamiltonian
+        coefficients :math:`c_i`.
+
+        Args:
+            coeffs (Sequence[float]): the coefficients of the terms in the Hamiltonian
+
+        Raises:
+            ValueError: if the learning rate is large than :math:`2/\sum |c_i|`
+        """
+        self.lipschitz = np.sum(np.abs(coeffs))
+
+        if self._stepsize > 2 / self.lipschitz:
+            raise ValueError(f"The learning rate must be less than {2 / self.lipschitz}")
+
+    def _single_shot_expval_gradients(self, expval_cost, args, kwargs):
+        """Compute the single shot gradients of an ExpvalCost object"""
+
+        qnodes = expval_cost.qnodes
+        coeffs = expval_cost.hamiltonian.coeffs
+        device = qnodes[0].device
+
+        self.check_device(device)
+        original_shots = device.shots
+
+        if self.lipschitz is None:
+            self.check_learning_rate(coeffs)
+
+        try:
+            if self.wrs:
+                grads = self.weighted_random_sampling(
+                    qnodes, coeffs, self.max_shots, self.trainable_args, *args, **kwargs
+                )
+            else:
+                device.shots = [(1, self.max_shots)]
+                grads = [
+                    qml.jacobian(expval_cost, argnum=i)(*args, **kwargs)
+                    for i in self.trainable_args
+                ]
+        finally:
+            device.shots = original_shots
+
+        return grads
+
+    def _single_shot_qnode_gradients(self, qnode, args, kwargs):
+        """Compute the single shot gradients of a QNode."""
+        device = qnode.device
+
+        self.check_device(qnode.device)
+        original_shots = device.shots

```

Comment:
See above: could we not use any function, but accept the device as an arg? Or is this a problem because we replace the device internally sometimes?

Reply:
>Or is this a problem because we replace the device internally sometimes?

Oh, I actually didn't think of that 😆 I didn't add the device as an arg due to my comments in https://github.com/PennyLaneAI/pennylane/pull/1139/files#r595922614, but I think this is maybe another reason not to add it as an arg.